### PR TITLE
Fix wrong tag and element size

### PR DESCRIPTION
**Pull request by ndorigatti**
There is a tag that is incorrect, looking at referenced links and doing real testing, this commit provides the fixes.

### CODE REVIEW

**Comment by kormax** [comment]
Hello. Thanks for your help.

Do you have any idea what this tag actually does? I've found "filter" in one of the device config manuals, but didn't find any side effects of using it in practice, apart from the following errors with some configs (for 9f2b by the way):
```
  sw1=0x67; sw2=0x00 if len > 5             
  sw1=0x6a; sw2=0x83 if 02 00000000                   
  sw1=0x62; sw2=0x87 if already read but 01 0000000   
  sw1=0x6a; sw2=0x80 if 00 00000000                     
```

Are you sure it's the same thing here? I could believe that they've made a mistake by putting B instead of 8 there, which could be a reason why my info was incorrect and I found no interesting side effects.

Looking forward to your response. Thanks.

**Comment by kormax** [comment]
I've just done some tests with tags, and have following findings:

1. Tag 9f2b: read is successful only if the first byte is 0x01 and length is 4 or 5 bytes. Otherwise, length errors or other random protocol-related errors.
2. Tag 9f28: read is successful if length is 4. Value can be any, doesn't impact operation at all.
3. Tag 9f2a (non-existent, to observe how it reacts to fake tags). Any length is accepted, any value is OK.

Judging from this info, I agree that 9f28 may indeed have some effect attached with it and it is worth noting, but without extra information I cannot be sure that it is indeed a filter (warrants "a fix"), and not something else.

**Comment by ndorigatti** [comment]
Hello, 
I can't currently test all the configurations you mention, but I'm sure I had to change 9F2b to 9F28, otherwise I could not get a correct response. And for the length, with 5 bytes i got error, with 4 (any byte) it was working.
I referred to one of your mentioned links:
https://gist.github.com/gm3197/ad0959476346cef69b75ea0523214350
There it states:
_Request TLV_

Tag | Length | Value | Notes
-- | -- | -- | --
9F22 | 2 | 0100 | Terminal Application Version
9F25 | 32 | "Merchant ID" | SHA-256 hash of Pass Type ID
9F28 | 4 | All values accepted | Required, nonce maybe?
9F26 | 4 | Terminal Capabilities | See table below
9F29 | Variable | Merchant URL | Optional

So I just followed these advices and made it read correctly. (for me *9F2* and not *2F2* as your last comment).
I did not do many tests as currently my goal was to read and decrypt successfully the pskt.io passes using an Android Device.

I'll dig more in the next months, for now I can only contribute with these little corrections and saying that 9F28 should be not important for now, but in my code I currently send the nonce received in the `select` command (tag 9F24).

9F2A is an unknown tag received in the getdata response (As far as i've seen!), see https://gist.github.com/gm3197/ad0959476346cef69b75ea0523214350#vas-response-tlv

9F2B is not in my request TLV currently, but reading your ocmment (first byte 01) maybe it is something like merchantID in Smart Tap?

Anyway, my current request contains these tags:
9F22 [02] 0100
9F25 -> Length + SHA256_of_passid
9F26 [04] 00800002
9F28 [04] (mobileNoncerecived or 00000000)
9F29 -> Lenght + Merchant URL

9F2a and 9F2b are right after 9F29 so that's why I think they could refer to merchant id or similar, as far as i can see the protocol is quite tidy, so probably first you have "versions", "hashes", "capabilities", "nonce/ephemeralkeys", "data params"

makes sense to you?
I did not found any other documentation except your site and the mentioned gist, so I'm just guessing. With Google Smart Tap is way easier!

**Comment by kormax** [comment]
Sorry for confusion, I've provided 9f2a as an example of a tag that's not used as an argument in any of the protocol commands. Forgot about 9f2a being in response, but it doesn't matter actually. The intent was to show that inexistent tags have no length and format requirements, unlike tags 9f28 and 9f2b, which means that they are indeed special.

Also, the "2f" in my post was a mistake, I've meant "9f" (going to fix that retroactively to remove confusion if anyone else reads this conversation). 
For reference, here's mine config (just added 9f28 as nonce, not sure what it is actually, that's why we're talking). 

```
tlvs = [
    # Version
    TLV(0x9f22, value=bytearray([0x01, 0x00])),
    # Pass identifier
    TLV(0x9f25, value=hashlib.sha256(pass_identifier.encode("ascii")).digest()),
    # Feature masks
    TLV(0x9f26, value=bytearray([0x00, terminal_info_mask, 0x00, capabilities_mask])),
    # Nonce, can be any
    TLV(0x9f28, value=bytes.fromhex("01020304")),
    # Filter, only accepts this format
    TLV(0x9f2b, value=bytes.fromhex("01 00000000")),
    # Merchant signup URL
    TLV(0x9f29, value=b'https://apple.com')      
]
```

With this config all of my reads are successful. I'm not sure why filter tag 9f2b caused any issues for you.

**Comment by ndorigatti** [comment]
Ok, now it's more clear to me! 
So you send both 9f28 and 9f2b... interesting, maybe if we add non existent tag it will work anyway? Like if we send 9f2f?
Maybe they reserved something for different "protocol" like paying or mixed content?
probably 9f2b gave me issues because I was sending a full 0 byte, no 01 at the start...

I used proxmark3 code to have some more help: https://github.com/RfidResearchGroup/proxmark3/blob/32f892e512d97be402c0abaf6f00675686ed0893/client/src/cmdhfvas.c#L121
They just send a 4bytes zero element and no 9f2b, no more ideas for me!

**Comment by kormax** [comment]
It does work if you add an inexistent tag (that's what I've meant by giving an example with 9f2a), but both 9f2b and 9f28 require a specific length, and 9f2b even needs a special format, so I'm sure that they actually defined in the protocol, we just don't know what they do. 

Regarding the use, you can omit tags 9f2b, 9f28, 9f29(url) and GET command should work anyway in VAS mode. Me and the other guy added them just for reference if someone wants to dig further.

**Comment by kormax** [comment]
If you don't have anything more to add regarding this finding:

I'll add the PR if you return the "filter" back, adding this new tag as "nonce" somewhere near it. I think it's worth keeping info about both for now until we find out the meaning.

**Comment by ndorigatti** [comment]
ok, should be filled now!

File: README.md
Comment (Mah8789) [inline README.md:1]: This is really helpful. But I still have one doubt, with these commands we are able to read the apple wallet, but what is the need for enhanced contactless polling.

**Pull request merged**